### PR TITLE
Handle ghc exceptions

### LINKED ISSUE #34: Unknown Language Extension causes crash on startup
Just need to catch another exception.

<details>
<summary>
Logs
</summary>

```
2019-10-14 18:08:11.464696271 [ThreadId 4] - Using plain GHC version
2019-10-14 18:08:11.511825829 [ThreadId 4] - Run entered for HIE(hie-8.6.4) Version 1.0.0.0 x86_64 ghc-8.6.4
2019-10-14 18:08:11.511958694 [ThreadId 4] - Current directory:/home/baldr/Documents/haskell/lambdabot-server
2019-10-14 18:08:11.578019101 [ThreadId 10] - Using plain GHC version
Using hie version: Version 1.0.0.0 x86_64 ghc-8.6.4
2019-10-14 18:08:11.652328132 [ThreadId 12] - New cradle: /home/baldr/Documents/haskell/lambdabot-server/src/Server/Utils.hs
2019-10-14 18:08:11.652740923 [ThreadId 12] - Found cradle: Cradle {cradleRootDir = "/home/baldr/Documents/haskell/lambdabot-server", cradleOptsProg = CradleAction: cabal}
Using hoogle db at: /home/baldr/.hoogle/default-haskell-5.0.17.hoo
hie-8.6.4: Unsupported extension: RecordWildcards
Perhaps you meant ‘RecordWildCards’ or ‘NoRecordWildCards’
```
</details>

### PR DESCRIPTION
**Pull request by jneira**
* Fixes #34
* Completes a subtask in #48
* Branch from 924de7 (cause hie-bios branch head is broken for me)

### CODE REVIEW

File: hie-plugin-api/Haskell/Ide/Engine/ModuleCache.hs [hunk 1]
```diff
@@ -182,8 +182,11 @@ loadCradle iniDynFlags (NewCradle fp) def action = do
         liftIO (GHC.newHscEnv iniDynFlags) >>= GHC.setSession
         liftIO $ setCurrentDirectory (BIOS.cradleRootDir cradle)
         init_res <- gcatches (Right <$> init_session)
-                [ErrorHandler (\(ex :: GHC.GhcException)
-                  -> return $ Left (GHC.showGhcException ex ""))]
+                [ ErrorHandler (\(ex :: AsyncException)
+                   -> liftIO (throwIO ex))
+                , ErrorHandler (\(ex :: SomeException)
+                   -> return $ Left (show ex))
+                ]
```
Comment (fendor) [inline hie-plugin-api/Haskell/Ide/Engine/ModuleCache.hs]: Eh, still opposed to handling Async Exceptions. I dont think we should try to catch all exceptions. In some cases, we just cant continue and crashing is the most sensible option.

Comment (jneira) [reply hie-plugin-api/Haskell/Ide/Engine/ModuleCache.hs]: Ok, so i will catch only `SourceError` (and `IOException`?)

Comment (fendor) [reply hie-plugin-api/Haskell/Ide/Engine/ModuleCache.hs]: I think, I would prefer that. What do @mpickering, @alanz, etc.. think?

Comment (alanz) [reply hie-plugin-api/Haskell/Ide/Engine/ModuleCache.hs]: Agree, rather build up a list of the ones we know it is safe to deal with. Which we had (maybe still do?) on the handler for the call to GHC when compiling.

Comment (jneira) [reply hie-plugin-api/Haskell/Ide/Engine/ModuleCache.hs]: Thanks for the tip @alanz: i found it and it seems to be just what we needed here: https://github.com/mpickering/haskell-ide-engine/blob/hie-bios/hie-plugin-api/Haskell/Ide/Engine/Ghc.hs#L203

**Comment by jneira** [comment]
Ok i've changed the code to use [`errorHandlers`](https://github.com/mpickering/haskell-ide-engine/blob/hie-bios/hie-plugin-api/Haskell/Ide/Engine/Ghc.hs#L203) like `Haskell.Ide.Engine.Ghc` already did.

The solution for `SourceError`s is the simpler one (but i am not sure if it is the *correct* one): let the cradle be initialized in case there are `SourceError`s cause hie will use it for compute diagnostics (that is the specialized place to do it) and report them for the user.

File: hie-plugin-api/Haskell/Ide/Engine/ModuleCache.hs [hunk 2]
```diff
@@ -181,12 +181,18 @@ loadCradle iniDynFlags (NewCradle fp) def action = do
         -- while 'f' is still valid.
         liftIO (GHC.newHscEnv iniDynFlags) >>= GHC.setSession
         liftIO $ setCurrentDirectory (BIOS.cradleRootDir cradle)
-        init_res <- gcatches (Right <$> init_session)
-                [ErrorHandler (\(ex :: GHC.GhcException)
-                  -> return $ Left (GHC.showGhcException ex ""))]
+
+        let onGhcError = return . Left
+        let onSourceError = const . return $ Right ()
```
Comment (jneira) [inline hie-plugin-api/Haskell/Ide/Engine/ModuleCache.hs]: Maybe we should log the `SourceError` although we continue setting the Cradle

Comment (jneira) [reply hie-plugin-api/Haskell/Ide/Engine/ModuleCache.hs]: Done, i've log the final setting of the cradle as a final sanity check

**Pull request merged**